### ISSUE TITLE
add default branches that are ignore by smithy for uploads

### ISSUE THREAD

**Issue by beckermr** [conda-forge/conda-smithy#1904]
Right now, one can set an appropriate key in the conda-forge.yml to have smithy only do uploads for certain branches. This is useful, but @jaimergp and I discussed a separate scheme that we think could be a nice default and enable some new things. The idea is to always disallow uploads from anything that matches the following globs/regex:

 - `_*` (i.e., any branch that starts with an underscore)
 - `^[^\S].-patch-[0-9].$` (i.e., any branch that matches the pattern of default GH branch names for edits in the UI, `<username>-patch-<number>`.
 - any branch that matches the default branch name for PR reversions (need to look this up)

If we moved to this system, then we'd avoid common user errors of uploads happening when they do not  fork the feedstock and we could use branches that start with `_` for automated tasks like rerendering.

Thoughts @conda-forge/core?

**Comment by h-vetinari** [comment]
Why not do an allowlist instead of a blocklist? Something like `main|dev|rc` + `^v?[\d\.]+(\.x)?$`; or maybe just `main` and enable extending that list through `conda-forge.yml`?

In any case, an allowlist seems way more waterproof than chasing holes in our blocklists to me.

**Comment by beckermr** [comment]
I don't care how it is phrased so happy with whatever in terms of allow vs reject lists. 

One vote in favor of an expanded reject list is that if we instated allow lists we'd probably break existing feedstocks at some point. 

The big point here, however, is that our current default of allowing everything is not as useful as several other possible default choices and I think we should change it.

**Comment by jaimergp** [comment]
> Adapting the relevant part of [this comment](https://github.com/conda-forge/conda-smithy/pull/2682#issuecomment-5669937187) here because it somehow became the plan:

The way I thought about doing this was to:

- [ ] Create an `admin-migrations` migration that traversed all feedstocks and annotated the branch name in each non-`main`'s `conda-forge.yml`'s `upload_on_branch` field.
  - [ ] https://github.com/conda-forge/admin-migrations/pull/221
- [ ] Once migrated, swap the default for `upload_on_branch` to `main` only.
  - [ ] https://github.com/conda-forge/conda-smithy/pull/2683
  - [ ] https://github.com/conda-forge/conda-smithy/pull/2684
- [ ] Update documentation to mention that any new branches after this need to explicitly set `upload_on_branch` and rerender, although the linter should already help a lot here.
- [ ] We could also add a conda-forge-admin command to create LTS branches more easily.

Thanks @beckermr for working on this!

**Comment by h-vetinari** [comment]
One aspect I don't 100% understand is why we have to keep the `upload_on_branches` settings per branch. I'd much prefer this to be taken exclusively from the main branch, for a few reasons:
* We already do the same for `abi_migration_branches`
* It's more centralised for ease of review
* The main branch is the only one that's protected by default

It would be especially confusing/annoying to have to add a new LTS branch in two different places (under `abi_migration_branches` on main, and under `upload_on_branches` on the new branch itself).

Speaking of which, the distinction between these two options seems to be almost completely negligible. Is there really a use-case for a branch that wants to upload packages, but doesn't want to receive migration PRs? Of course, a dev branch may not *need* those, but it'd be easy to ignore the occasional bot PR (already a reality for LTS branches anyway, though much improved by recent "has this been migrated already?" detection).

What I'm getting at is that we may want to unify `abi_migration_branches` and `upload_on_branches`, rather than force maintainer to duplicate this configuration in 99% of cases.

**Comment by beckermr** [comment]
The bot works differently because it only parses the default branch recipe to collect its metadata. 

The setting needs to be per branch because of how the ci support code works. That code both checks the setting and parses the recipe on the branch. If that code had to checkout the main branch to check the setting, there would be a whole host of issues and/or non-ideal changes. These include

1. We'd need non-shallow checkouts of the feedstock or to separately pull the the feedstock at the default branch to do the check. 
2. Uploads from the branch would be non-deterministic since they would depend on the state of the main branch whenever the upload happens. Literally some packages from a ci run could upload, but then the default branch could change, and so the rest could fail. 
3. Folks would need to make prs onto multiple branches to enable uploads from a new branch. They'd need a pr into the main branch to turn on uploads in addition to any work / prs needed for the branch itself.


Using upload on branch creates a very easy workflow. To upload from a new branch, you

1. Make that branch. 
2. Make a pr targeting that branch from a different new branch which makes any changes to recipe and sets upload on branch to the branches name from 1. 
3. Merge the pr once it passes. 

Since upload on branch in the pr branch in 2 doesn't match the name of the branch from 1, uploads are skipped in the pr process. 

This mismatch causing uploads to fail is the key bit that makes things intuitive. Any new branch won't get uploads, and so folks can simply make branches as they please.  With a centralized list, people will eventually make a branch that happens to match the list and then get uploads when they did not intend to.

**Comment by beckermr** [comment]
I'm going to move on from this task now. If there is another system you prefer, feel free to pick up the work here and finish it.

**Comment by beckermr** [comment]
Another way to think about this is that it is actually the bot setting that is confusing. 

Changing the bot settings to work per branch would be a better approach to unify things since it could provide a centralized oath to enable other interesting things like version prs for patch versions onto version branches.

**Comment by beckermr** [comment]
> With a centralized list, people will eventually make a branch that happens to match the list and then get uploads when they did not intend to.

For example, in the version branch from globs example, this would happen basically immediately. If a centralized list has a glob that matches any version tag of the form `x.y.z`, then as soon as someone wants to add a new version branch, that branch immediately gets uploads when it is made. This is the whole problem we are trying to solve with the allow list.

**Comment by h-vetinari** [comment]
Thanks for the responses!

> I'm going to move on from this task now. If there is another system you prefer, feel free to pick up the work here and finish it.

I don't understand why it's seemingly a big ask(?) to do some minimal design work before we decide on the shape of a new feature that will likely not get touched for a few years. Doing this kind of discussion in advance is time well-spent. And yes, that may include challenging certain assumptions, and whether the costs they impose merit a different approach. This should be par for the course, not some kind of outlandish demand to be granted reluctantly. I'm not saying we should turn in circles on this forever, just be able to discuss it serenely at least once.

> We'd need non-shallow checkouts of the feedstock or to separately pull the the feedstock at the default branch to do the check.

It would be sufficient to query `https://raw.githubusercontent.com/conda-forge/<package>-feedstock/refs/heads/main/conda-forge.yml`, entirely without a checkout, much less a non-shallow one. 

> Uploads from the branch would be non-deterministic since they would depend on the state of the main branch whenever the upload happens. Literally some packages from a ci run could upload, but then the default branch could change, and so the rest could fail.

That sounds completely fine? `main` has the reference data. If it doesn't contain the branch you want to upload from - too bad, go fix it and rerun CI.

> Folks would need to make prs onto multiple branches to enable uploads from a new branch. They'd need a pr into the main branch to turn on uploads in addition to any work / prs needed for the branch itself.

That's already the case for adding a new LTS branch under your proposed scheme. People would have to push a change to `main` for `abi_migration_branches`, and one to the branch itself for `upload_on_branch`. My preferred approach would be: enable _once_ on `main`, and then raise a PR as usual to the respective branch (and also benefit from relevant migration PRs henceforth, without yet more duplicated config).

> This mismatch causing uploads to fail is the key bit that makes things intuitive. Any new branch won't get uploads, and so folks can simply make branches as they please. With a centralized list, people will eventually make a branch that happens to match the list and then get uploads when they did not intend to.

That scenario does not seem likely to me. For one, it assumes that people make PRs from branches on the repo itself, rather than from a fork, which we've heavily discouraged (and I think should continue to do so). And then, if `conda-forge.yml` looks like
```yaml
branches:  # just for illustration: unified config between abi_migration_branches and upload_on_branches
  - main
  - v20.x
  - v19.x
```
it's not like people will randomly fall into the same names. Besides, only maintainers can push branches directly, and those maintainers should be aware of their own feedstocks' branch configuration (especially which ones cause uploads).

> If a centralized list has a glob that matches any version tag of the form `x.y.z`, then as soon as someone wants to add a new version branch, that branch immediately gets uploads when it is made. This is the whole problem we are trying to solve with the allow list.

We don't _have_ to allow globs; I'd prefer explicitly listing the maintenance branches TBH. Also: see above how the only people who can push branches directly are also the ones who need to know the branch setup. In other words, the exact same concern also appears in your model, for a maintainer pushing a random `x.y.z` branch that happens to match the glob that's already contained in the respective `conda-forge.yml`. To me that's an argument against globbing, but unrelated to the question on which branch we save the reference data.

> Changing the bot settings to work per branch would be a better approach to unify things since it could provide a centralized path to enable other interesting things like version prs for patch versions onto version branches.

I'm not sure I see the relation between using per-branch bot config and being able to issue version PRs for those branches. Could you elaborate? I still like the fact that there's a single place to look for `abi_migration_branches`, and that I don't have to go hunting through old branches, and whether they might still be "live".

**Comment by beckermr** [comment]
> https://raw.githubusercontent.com/conda-forge/<package>-feedstock/refs/heads/main/conda-forge.yml

GitHub has started rate-limiting these requests, which has caused transient failures in other parts of our infrastructure that use raw URLs. I would not want to introduce that here too.

It's my time that is the limit here, and that's why I need to move on. As of now, the task is basically done, besides a few PR comments that would take a few minutes. With the plan from @jaimergp, we don't need any changes to the ci setup package, and so there are no more PRs to be made. I picked this task up because it was quick and high impact. When it goes from quick to not quick, I need to bow out.

**Comment by h-vetinari** [comment]
I think there are multiple distinct questions here, and mixing them makes finding a solution here harder, so let me try to separate them

* should we switch to an opt-in model for which branches can upload packages?
  * absolutely! It's good security posture, and also necessary for some automated workflows that can only run in branches directly on the feedstock
* should we keep distinct options `upload_on_branches` and `abi_migration_branches`?
  * IMO no, once we switch to requiring `upload_on_branches`, the distinction between them becomes so miniscule as to be irrelevant
* whether unified or not, should  `upload_on_branches` and `abi_migration_branches` live in the same place?
  * absolutely! It's a big footgun for maintainers to have one in one place (`upload_on_branches` on the respective branch) and one in another (`abi_migration_branches` on `main`)
* assuming we want to unify the location, should it be on the respective branch or on main?
  * IMO on main, though the reasons are more nice-to-have than critical (e.g. have all "live" branch config easily findable in a single place, rather than having to check N branches), and so both options are feasible
* how do we achieve the actual implementation, if we want to take both cases from one place?
  * You're more familiar with the constraints there. Happy to defer to your expertise, e.g. in avoiding requests. FWIW, I think a shallow clone would be enough to get the content of a `conda-forge.yml` cheaply. Other options may include persisting the relevant information somewhere else that we control (like in the graph?).
* Do we want to allow globs for `upload_on_branches`?
  * IMO no, the downsides outweigh the benefits.
* Do we want users to more generally switch to workflows based on branches directly on the feedstock, rather than forks
  * IMO no, but that's mostly unrelated here

**Comment by jaimergp** [comment]
One thing to note:

- `conda-forge.yml` configures `conda-smithy`, _and_ happens to be the place where we place the `bot:` config which gets queried by the bot to see what to do. 
- In my opinion, these two things should be decoupled, since smithy and the rendered CI can run without the bot existing at all.

So, if we forget that `conda-forge.yml` happens to contain bot configuration, it makes sense to have the configuration isolated per-branch. After all, that's what happens with the recipe and conda_build_config.yaml. I'll repeat that: `conda-forge.yml` is the smithy configuration file to rerender a repository branch, and that results in self-contained configuration that doesn't get resolved at runtime. More deterministic too, less surface to network issues.

Anything that is global to the repo should probably live in a different file to avoid confusion. So, if anything, I'd support moving the `bot:` section to separate file that happens to live in `main` out of convenience.

So, in that sense, I reject the proposal to use `upload_branches` as a list of branches. It goes against the notion of "this file is branch configuration". That's why I proposed that plan above, sticking to `upload_on_branch`.

**Comment by beckermr** [comment]
Yes, this is the detailed version of my point that I made somewhere about it actually being the bot settings that are weird here.

**Comment by h-vetinari** [comment]
> In my opinion, these two things should be decoupled, since smithy and the rendered CI can run without the bot existing at all.

That's a fair position to take, and one I largely agree with; which branches are upload-enabled is not a branch-specific choice though (except perhaps by ease of implementation), but rather a repo-wide policy. This is especially so IMO because we already have another place (separately as you say, for the bot), where we need to carry that information with 99% overlap for `abi_migration_branches`.

> So, in that sense, I reject the proposal to use `upload_branches` as a list of branches. It goes against the notion of "this file is branch configuration".

We can reject each other's inputs all day long, but that doesn't get us anywhere -- you offer no solution to the "duplicate config in separate places" concern. I'm open to pretty much any approach that resolves this.

How about we:
* split out the repo-level config from `conda-forge.yml` and put the `bot:` config in there
* agree that upload policy is a repo-level decision, rather than a branch-level one (for auditability, unification, convenience[^1], etc.)
* find a way to implement this & roll it out

Note that the repo-level decisions could also be kept outside of the feedstock, similar to how the output validation is already stored outside the feedstocks as well. IOW, we have room for creativity where we store this information, and how to interact with it (e.g. through admin-requests).

Piggybacking on what we have for output validation seems to be a promising option to me for this?

[^1]: maintenance branches across loads of feedstocks are full of stale configuration for `abi_migration_branches`. The same would happen with `upload_on_branches`, except that the latter then becomes a live grenade for blowing your foot off.

**Comment by jaimergp** [comment]
> which branches are upload-enabled is not a branch-specific choice though (except perhaps by ease of implementation), but rather a repo-wide policy.

The way I see it, whether we upload or not on a given branch is a boolean attached to that particular branch.

> you offer no solution to the "duplicate config in separate places" concern.

Because I don't see it as a concern. Yes, it's a bit meh to have the list of branches there too, but I don't see it as enough of a pain to justify the refactor and changed mental model of what a rerender entails.

---

> How about we:
> 
> 1. split out the repo-level config from conda-forge.yml and put the bot: config in there
> 2. agree that upload policy is a repo-level decision, rather than a branch-level one (for auditability, unification, convenience[1](https://github.com/conda-forge/conda-smithy/issues/1904#user-content-fn-1-9d57e06c3f408dd9e2d991b90b7c6c5e), etc.)
> 3. find a way to implement this & roll it out


1. We can do that already but I don't think we need to do it for this issue. And it also needs changes in the bot infra and a roll out. The action item here is to create a separate issue for that.
2. I disagree with this point of view.
3. And as a result the available PRs opened by Matt are enough to cover this transition.

The point driving all this is that it will make security concerns and token management in the automated PR submission infrastructure WAY simpler. That's what I'm interested in right now. We can help users understand how the configuration is setup and why with documentation and linter rules.

**Comment by beckermr** [comment]
I am going to pile on here against my better judgment. 

By moving the bot to a per-branch configuration, we can repurpose the current version ignore fields, etc., to drive per-branch patch/micro/minor version updates, and we can transform the bot ABI migration branches field into a single boolean.

The change on the bot end is to glob config info across all branches and to adjust the version migrator to pull recipes from the branch it is producing a PR for. 

Focusing effort there is much more interesting to me personally.

**Comment by h-vetinari** [comment]
> By moving the bot to a per-branch configuration, we can repurpose the current version ignore fields, etc., to drive per-branch patch/micro/minor version updates, and we can transform the bot ABI migration branches field into a single boolean.

Ironically, moving everything into one place addresses my main concern.

> The way I see it, whether we upload or not on a given branch is a boolean attached to that particular branch.

Having a "can_this_branch_upload_packages?" boolean, combined with a setup where maintainers & bots are free (or even encouraged) to push to the feedstock, is directly at odds with the stated goal of avoiding accidental uploads. Say someone wants to raise a PR to a maintenance branch, and branches off from the current HEAD of that branch; if they don't also remember to clean up the boolean before pushing their work, this _will_ cause such accidental uploads.

In other words, this aspect of the design is comically bad. It's baffling to me why you're so keen to inflict this on maintainers, except perhaps due to a kind of "quick, quick, we have to use this square wheel, because we don't have time to get a round one" mentality.

But I'm done trying to work out something better, seeing that you're determined to dismiss my concerns or ideas (like handling this along the existing infrastructure for gating uploads due to output validation) out of hand.

**Comment by beckermr** [comment]
I think maybe one aspect here is misunderstood.

`upload_on_branch` is not a boolean. It is a string specifying the branch name that allows uploads for the given `conda-forge.yml` configuration. I think of it as the "target upload branch" for the `conda-forge.yml` or maybe the `conda-forge.yml`'s "home branch."

I have been trying to think of a clearer name for this parameter, but I have not come up with anything. `target_upload_branch` makes sense to me, but it is hard to say if that is clearer.

When `upload_on_branch` is set, uploads happen only when the `upload_on_branch` value in the `conda-forge.yml` matches the branch name. (Uploads also currently happen if `upload_on_branch` is not set at all. This is the behavior that @jaimergp is suggesting we deprecate and then remove by making the default value be the default branch name.) See the code [here](https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/f7bd98b4cfe77fbfbbb4af1741ba1f7ab9eec60b/recipe/conda_forge_ci_setup/build_utils.py#L224-L234).

So as soon as you make a new branch off of any branch where uploads work with `upload_on_branch` set, then uploads won't happen for the new branch since the new branch doesn't have the same name, but has the same `conda-forge.yml` file contents. 

This means that a valid workflow for creating a new version/LTS branch without extra uploads is:

1. Make the version/LTS branch off of w/e branch you please.
2. Make another branch off of the version/LTS branch for the PR head.
3. Do your work in the PR head branch to build the version
4. Make sure to enable uploads by setting `upload_on_branch` in the PR head branch to the version/LTS branch name.
5. Merge the PR head branch into the version/LTS branch.

It also means that any branch made on the feedstock will not cause uploads unless a commit is made to the branch.

**Comment by beckermr** [comment]
On the topic of storing the upload information outside of the feedstock (in eg feedstock-outputs), I am not a huge fan of that. The only reason we store stuff in feedstock-outputs is for security. If we could avoid that extra repo plus keeping it synced, we'd definitely do that. So I don't think it is a good idea to put this config info in another spot besides the feedstock.

**Comment by beckermr** [comment]
I've been quietly using a version of this workflow on the repodata patches feedstock for a while. Upload on branch is set to main there.
